import emoji async draft

rinej opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

**Problem**
Currently, both translations for emoji in English (EN) and Spanish (ES) are imported during app initialization. Additionally, we create large objects of EmojiTrie for both languages. The idea is to import and build EmojiTrie only for the currently selected language.

**Solution**
During app initialization, we check the preferred language and import only that translation file.
We create an EmojiTrie only for that selected language.
If the user changes the language, we will dynamicly import the new translation file and create a new Trie object at that time.

In most cases, users only use one language, so by implementing this change, we can significantly reduce memory consumption and the time needed for creating large EmojiTrie objects.

**Some concerns**
Jest doesn't support dynamic imports by default. We have to add the flag `--experimental-vm-modules` (https://jestjs.io/docs/ecmascript-modules ) when running jest command (both in GH action and in local package json).

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/41893



### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->



### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- Open report view
- Type some emoji shortcut in the input field e.g. `:snail`
- See that the Snail emoji is in suggested list
- Change language in settings to spanish
- Check if translation are loaded properly
- Go back to the report screen
- Type spanish emoji shortcut e.g. :caracol
- See that the Snail emoji is in suggested list

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/2650904/e5813406-ce5f-4104-a600-c21dfd30ed84


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/2650904/cd85802d-8edf-4cf6-8322-a1210d2e9d56


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/2650904/d42f6319-ea7b-486b-a1a9-9c392f5f121e


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/2650904/c8f41857-2cc6-4c93-8e29-fa62093f7097


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/2650904/b7c2fa3c-fbfd-46b0-9b0c-1795fb03f747


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/2650904/c5ab4979-bda0-4dc8-b7b7-deb3714ac0a4


<!-- add screenshots or videos here -->

</details>

rinej committed:
import emoji async draft

gedu commented on src/libs/EmojiTrie.ts:
```diff
@@ -95,9 +95,24 @@ function createTrie(lang: SupportedLanguage = CONST.LOCALES.DEFAULT): Trie<Emoji
 }
 
-const emojiTrie: EmojiTrie = supportedLanguages.reduce((prev, cur) => ({...prev, [cur]: createTrie(cur)}), {});
+// const emojiTrie: EmojiTrie = supportedLanguages.reduce((prev, cur) => ({...prev, [cur]: createTrie(cur)}), {});
```
Remove this comment when the PR is ready

rinej committed:
refactor: remove redundant if statements

hurali97 commented on src/libs/EmojiTrie.ts:
```diff
@@ -95,9 +96,22 @@ function createTrie(lang: SupportedLanguage = CONST.LOCALES.DEFAULT): Trie<Emoji
+const emojiTrie: EmojiTrie = {
+    en: undefined,
+    es: undefined,
+};
```
Instead of manually creating and maintaining this variable can we consider implementing something like:

https://github.com/callstack-internal/Expensify-App/blob/f903144c44c252e676670ad8d8251f992219cfd6/src/libs/Localize/index.ts#L68-L73

gedu replied:
That approach I think will create all the tries for "en" and "es" and we want to create them on demand, that's why the idea is to create them manually when it is needed

rinej replied:
good point, I added the change so now we create it based on `supportedLanguages`  variable, thanks!

hurali97 commented on src/libs/EmojiTrie.ts:
```diff
@@ -95,9 +96,22 @@ function createTrie(lang: SupportedLanguage = CONST.LOCALES.DEFAULT): Trie<Emoji
+    es: undefined,
+};
+
+const buildEmojisTrie = (locale: 'en' | 'es') => {
```
Based on the above comment we might need to tweak the locale type.

adhorodyski replied:
```ts
const locales = ['en', 'es'] as const;

type Locale = (typeof locales)[number]; // produces a type of: 'en' | 'es';
```

adhorodyski replied:
cc @rinej

rinej replied:
adjusted it, used Locale type, thanks!

rinej committed:
add dynamic emojiTrie creation based on locale variable

rinej committed:
Merge branch 'main' into Load-Emojis-async-depending-on-lang

rinej committed:
use vm module for jest tests

rinej committed:
add proper error handling for dynamic imports

rinej committed:
fix: fix lint

rinej committed:
add console logs for debug on cli

rinej committed:
add build stept to test to debug on cli

rinej committed:
fix lint

rinej committed:
Merge branch 'main' into Load-Emojis-async-depending-on-lang

rinej committed:
remove silent and add logs for debug

rinej committed:
add cl fo debug

rinej committed:
run test action from packagejson script

rinej committed:
add experimental flag to gh action

rinej committed:
adjust action

rinej committed:
adjust jest action

rinej committed:
cleanup

rinej committed:
add comments

rinej committed:
cleanup

rinej committed:
adjust package json script

rinej committed:
adjust imports for ios

rinej committed:
parse typescript types for jest tests

rinej committed:
Merge branch 'main' into Load-Emojis-async-depending-on-lang

rlinoz reviewed (dismissed):
Looks good to me!

rlinoz commented:
All yours @mountiny

dubielzyk-expensify commented:
Looks like there's no design review needed here as this is mostly a technical solution?

mountiny reviewed (changes requested):
@rinej Can you please merge main and check out the failing test? I am worried it might be related to the node_options change?

Also can you please update the QA steps to not be a checkbox, but instead a numbered list? Its easier to read and follow. Thank you!

rinej committed:
Merge branch 'main' into Load-Emojis-async-depending-on-lang

rinej committed:
fix failing workflow

rinej commented:
@mouting I did it, I also added the change to fix the failing workflow

There was similar issue here -> https://github.com/Expensify/App/pull/40778#issuecomment-2096382864
I adjusted the tests suggested here -> https://github.com/Expensify/App/pull/40778/commits/ccfd44ba8a5c7b120f499eb8c76df6d3cf48efc0

mountiny reviewed (approved):
Thanks for the changes

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9093471283) to staging by https://github.com/mountiny in version: 1.4.74-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9199894178) to production by https://github.com/chiragsalian in version: 1.4.74-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
